mlir/include/mlir/Dialect/Utils/ReshapeOpsUtils.h

```diff
@@ -305,8 +306,42 @@ struct ComposeCollapseOfExpandOp : public OpRewritePattern<CollapseOpTy> {
       rewriter.replaceOpWithNewOp<CollapseOpTy>(
           collapseOp, resultType, expandOp.getSrc(), composedReassociation);
     } else if (srcRank < resultRank) {
+      // Compute the dynamic output shape for the new expand_shape op.
+      Location loc = collapseOp.getLoc();
+      SmallVector<OpFoldResult> origOutputShape =
+          expandOp.getMixedOutputShape();
+      SmallVector<OpFoldResult> newOutputShape;
+      for (auto indices : collapseOp.getReassociationIndices()) {
+        int64_t numStaticElems = 1;
+        SmallVector<Value> dynamicSizes;
+        for (auto idx : indices) {
+          OpFoldResult size = origOutputShape[idx];
+          if (auto maybeCst = getConstantIntValue(size)) {
```

Comment:
nit: spell out types
```suggestion
      for (SmallVector<int64_t, 2> indices : collapseOp.getReassociationIndices()) {
        int64_t numStaticElems = 1;
        SmallVector<Value> dynamicSizes;
        for (int64_t idx : indices) {
          OpFoldResult size = origOutputShape[idx];
          if (std::optional<int64_t> maybeCst = getConstantIntValue(size)) {
```
(don't remember the exact type for reassociation indices off the top of my head)